go/libraries/doltcore/diff/async_differ.go

```diff
@@ -91,13 +118,97 @@ func (ad *AsyncDiffer) GetDiffs(numDiffs int, timeout time.Duration) ([]*diff.Di
 	}
 }
 
-func (ad *AsyncDiffer) ReadAll() ([]*diff.Difference, error) {
-	diffs, hasMore, err := ad.GetDiffs(0, 5*time.Minute)
-	if err != nil {
-		return nil, err
+type keylessDiffer struct {
+	*AsyncDiffer
+
+	df         diff.Difference
+	copiesLeft uint64
+}
+
+var _ RowDiffer = &keylessDiffer{}
+
+func (kd *keylessDiffer) GetDiffs(numDiffs int, timeout time.Duration) (diffs []*diff.Difference, more bool, err error) {
+	timeoutChan := time.After(timeout)
+	diffs = make([]*diff.Difference, numDiffs)
+	idx := 0
+
+	for {
+		// first populate |diffs| with copies of |kd.df|
+		for (idx < numDiffs) && (kd.copiesLeft > 0) {
+			diffs[idx] = &kd.df
+
+			idx++
+			kd.copiesLeft--
+		}
+		if idx == numDiffs {
+			return diffs, true, nil
+		}
+
+		// then get another Difference
+		var d diff.Difference
+		select {
+		case <-timeoutChan:
+			return diffs, true, nil
+
+		case <-kd.egCtx.Done():
+			return nil, false, kd.eg.Wait()
+
+		case d, more = <-kd.diffChan:
+			if !more {
+				return diffs[:idx], more, nil
+			}
+
+			kd.df, kd.copiesLeft, err = convertDiff(d)
+			if err != nil {
+				return nil, false, err
+			}
+		}
 	}
-	if hasMore {
-		return nil, errors.New("Unable to read the diffs in a reasonable amount of time")
+
+}
+
+// convertDiff reports the cardinality of a change,
+// and converts updates to adds or deletes
+func convertDiff(df diff.Difference) (diff.Difference, uint64, error) {
+	var oldCard uint64
+	if df.OldValue != nil {
+		v, err := df.OldValue.(types.Tuple).Get(row.KeylessCardinalityValIdx)
+		if err != nil {
+			return df, 0, err
+		}
+		oldCard = uint64(v.(types.Uint))
+	}
+
+	var newCard uint64
+	if df.NewValue != nil {
+		v, err := df.NewValue.(types.Tuple).Get(row.KeylessCardinalityValIdx)
+		if err != nil {
+			return df, 0, err
+		}
+		newCard = uint64(v.(types.Uint))
+	}
+
+	switch df.ChangeType {
+	case types.DiffChangeRemoved:
+		return df, oldCard, nil
+
+	case types.DiffChangeAdded:
+		return df, newCard, nil
+
+	case types.DiffChangeModified:
+		delta := int64(newCard) - int64(oldCard)
+		if delta > 0 {
+			df.ChangeType = types.DiffChangeAdded
+			df.OldValue = nil
+			return df, uint64(delta), nil
+		} else if delta < 0 {
+			df.ChangeType = types.DiffChangeRemoved
+			df.NewValue = nil
+			return df, uint64(-delta), nil
+		} else {
+			return df, 0, fmt.Errorf("diff with delta = 0 for key: %s", df.KeyValue.HumanReadableString())
```

Comment:
Should this be a panic? Should be impossible, right?

Reply:
Should be, yes...